File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
  constructor(realNumber, imaginaryNumber) {
      this._realNumber = realNumber;
      this._imaginaryNumber = imaginaryNumber;
  }

  get real() {
      return this._realNumber;
  }

  get imag() {
      return this._imaginaryNumber;
  }

  get abs() {
      return Math.sqrt(this._realNumber ** 2 + this._imaginaryNumber ** 2);
  }

  add(complexNumber) {
      return new ComplexNumber(this._realNumber + complexNumber.real, this._imaginaryNumber + complexNumber.imag);
  }

  sub(complexNumber) {
      return new ComplexNumber(this._realNumber - complexNumber.real, this._imaginaryNumber - complexNumber.imag);
  }

    div(complexNumber) {
        let denominator = (complexNumber.real ** 2 + complexNumber.imag ** 2);

        return new ComplexNumber((this._realNumber * complexNumber.real + this._imaginaryNumber * complexNumber.imag) / denominator,
            (this._imaginaryNumber * complexNumber.real - this._realNumber * complexNumber.imag) / denominator);
     
  }

  mul(complexNumber) {
      return new ComplexNumber(this._realNumber * complexNumber.real - this._imaginaryNumber * complexNumber.imag,
          this._imaginaryNumber * complexNumber.real + this._realNumber * complexNumber.imag);
  }

  get conj() {
      return new ComplexNumber(this._realNumber, this._imaginaryNumber === 0 ? 0 : -1 * this._imaginaryNumber);
  }

  get exp() {
      let factor = Math.exp(this._realNumber);

      return new ComplexNumber(factor * Math.cos(this._imaginaryNumber), factor * Math.sin(this._imaginaryNumber));
  }
}

File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double _real;
    private readonly double _imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        _real= real;
        _imaginary = imaginary;
    }

    public double Real()
    {
        return _real;
    }

    public double Imaginary()
    {
        return _imaginary;
    }

    public ComplexNumber Mul(ComplexNumber other)
    {
        var otherReal = other.Real();
        var otherImaginary = other.Imaginary();
        
        var real = _real * otherReal - _imaginary * otherImaginary;
        var imaginary = _imaginary * otherReal + _real * otherImaginary;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Mul(int factor)
    {
        var otherReal = factor;
        var otherImaginary = 0;
        
        var real = _real * otherReal - _imaginary * otherImaginary;
        var imaginary = _imaginary * otherReal + _real * otherImaginary;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Add(int factor)
    {
        var otherReal = factor;
        var otherImaginary = 0;

        var real = _real + otherReal;
        var imaginary = _imaginary + otherImaginary;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Add(ComplexNumber other)
    {
        var otherReal = other.Real();
        var otherImaginary = other.Imaginary();

        var real = _real + otherReal;
        var imaginary = _imaginary + otherImaginary;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Sub(ComplexNumber other)
    {
        var otherReal = other.Real();
        var otherImaginary = other.Imaginary();

        var real = _real - otherReal;
        var imaginary = _imaginary - otherImaginary;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Div(ComplexNumber other)
    {
        var otherReal = other.Real();
        var otherImaginary = other.Imaginary();

        var divisor = (Math.Pow(otherReal,2) + Math.Pow(otherImaginary,2));
        var real = (_real * otherReal + _imaginary * otherImaginary) / divisor;
        var imaginary = (_imaginary * otherReal - _real * otherImaginary) / divisor;

        return new ComplexNumber(real, imaginary);
    }

    public ComplexNumber Div(int factor)
    {
        var otherReal = factor;
        var otherImaginary = 0;

        var divisor = (Math.Pow(otherReal,2) + Math.Pow(otherImaginary,2));
        var real = (_real * otherReal + _imaginary * otherImaginary) / divisor;
        var imaginary = (_imaginary * otherReal - _real * otherImaginary) / divisor;

        return new ComplexNumber(real, imaginary);
    }

    public double Abs() => Math.Sqrt(Math.Pow(_real,2) + Math.Pow(_imaginary,2));
    
    public ComplexNumber Conjugate() => new ComplexNumber(_real, -_imaginary);
    
    public ComplexNumber Exp() => new ComplexNumber(Math.Exp(_real) * Math.Cos(_imaginary), Math.Exp(_real) * Math.Sin(_imaginary));
}